File: src/main/kotlin/domain/auth/SessionActions.kt
@file:OptIn(ExperimentalTime::class)

package com.kitledger.domain.auth

import com.kitledger.services.config.AppConfig
import com.kitledger.services.database.SessionsTable
import com.kitledger.services.utils.generateUuidV7
import org.jetbrains.exposed.v1.r2dbc.insert
import java.util.*
import kotlin.time.Clock
import kotlin.time.Duration.Companion.seconds
import kotlin.time.ExperimentalTime

/**
 * Starts a new session for a user.
 * @param userId the user ID.
 * @return the session, or null if it could not be started.
 */
suspend fun startSession(userId: UUID): Session? {

    try {
        val sessionConfig = AppConfig.sessionConfig
        val sessionId = generateUuidV7()
        val sessionResult = SessionsTable.insert {
            it[SessionsTable.id] = sessionId
            it[SessionsTable.userId] = userId
            it[createdAt] = Clock.System.now()
            it[expiresAt] = Clock.System.now().plus(sessionConfig.ttl.seconds)
        }

        val session = sessionResult.resultedValues?.get(0)?.toSession()

        return session
    } catch (e: Exception) {
        return null
    }
}

File: src/domain/auth/session_actions.ts
import { db } from "../../services/database/db.ts";
import { sessions } from "../../services/database/schema.ts";
import { sessionConfig } from "../../config.ts";
import {v7} from "uuid";

export async function startSession(userId: string): Promise<string> {
	const sessionId = v7();

	await db.insert(sessions).values({
		id: sessionId,
		user_id: userId,
		expires_at: new Date(Date.now() + (sessionConfig.ttl * 1000)),
		created_at: new Date(),
	});

	return sessionId;
}
